python/ray/data/datasource/datasink.py

```diff
@@ -0,0 +1,77 @@
+from typing import Any, Iterable, List
+
+from ray.data._internal.execution.interfaces import TaskContext
+from ray.data.block import Block
+
+WriteResult = Any
+
+
+class Datasink:
+    def on_write_start(self, **_) -> None:
+        """Callback for when a write job starts.
+
+        Use this method to perform setup for write tasks. For example, creating a
+        staging bucket in S3.
+
+        Args:
+            _: Forward-compatibility placeholder.
+        """
+        pass
+
+    def write(
+        self,
+        blocks: Iterable[Block],
+        ctx: TaskContext,
+        **_,
+    ) -> WriteResult:
+        """Write blocks. This is used by a single write task.
+
+        Args:
+            blocks: List of data blocks.
+            ctx: ``TaskContext`` for the write task.
+            _: Forward-compatibility placeholder.
+
+        Returns:
+            The output of the write task.
+        """
+        raise NotImplementedError
+
+    def on_write_complete(self, write_results: List[WriteResult], **_) -> None:
+        """Callback for when a write job completes.
+
+        This can be used to "commit" a write output. This method must
+        succeed prior to ``write_datasink()`` returning to the user. If this
+        method fails, then ``on_write_failed()`` is called.
+
+        Args:
+            write_results: The list of the write task results.
+            _: Forward-compatibility placeholder.
+        """
+        pass
+
+    def on_write_failed(self, error: Exception, **_) -> None:
```

Comment:
I think it'd be good to pass in the WriteResults of any tasks that did succeed. Another possibility is we pass a List[Optional[WriteResult]], one for each task index, and we store an exception(s) as the WriteResult for a failed task.

Reply:
I agree that it'd be helpful, but how would we access the results from the successful tasks? Currently, we implement writes with something like `dataset.map(write_fn)`. I wonder if there's a simple way to get the successful task results without breaking abstraction barriers or leaking too many implementation details?

https://github.com/ray-project/ray/blob/0ec69f0692fdbb6c3286b61ab3057ad9d5db2b52/python/ray/data/dataset.py#L3459-L3472

Reply:
I think if we use the streaming executor, it should be possible to pass the blocks that succeeded so far?

Reply:
Yeah, I'm sure it's possible. I guess I'm more curious about how we'd implement this without introducing a lot of complexity. 

The current write implementation uses high-level APIs. My initial impression is that you'd need to access several low-level abstractions in `Dataset.write_datasink` to get the successful tasks.

Reply:
After offline sync - since this is for internal use only, I think this is fine.